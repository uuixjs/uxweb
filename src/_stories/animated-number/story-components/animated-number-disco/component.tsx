import * as React from "react";

import { AnimatedNumber, Layout } from "v2";

import { Component } from "react";

interface DiscoProps {
  startValue: number;
  deltaMin?: number;
  deltaMax?: number;
}

interface DiscoState {
  value: number;
}

export class AnimatedNumberDisco extends Component<DiscoProps, DiscoState> {
  private interval: number | undefined;

  public constructor(props: DiscoProps) {
    super(props);
    this.state = {
      value: props.startValue,
    };
  }

  public componentDidMount() {
    this.interval = window.setInterval(this.tick, 2000);
  }

  public componentWillUnmount() {
    window.clearInterval(this.interval);
  }

  public render() {
    return (
      <Layout>
        <AnimatedNumber
          value={this.state.value}
          formatValue={(n) => n.toLocaleString()}
        />
      </Layout>
    );
  }

  private tick = () => {
    const diff = generateRandomInteger(
      this.props.deltaMin || -10,
      this.props.deltaMax || 10,
    );

    this.setState((prevState) => ({
      value: prevState.value + diff,
    }));
  };
}

function generateRandomInteger(min: number, max: number) {
  return Math.floor(min + Math.random() * (max + 1 - min));
}
