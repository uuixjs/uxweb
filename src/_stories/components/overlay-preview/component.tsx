import "./styles.scss";

import * as React from "react";

import {
  AlignItems,
  Background,
  CoreImage,
  Display,
  InjectLayout,
  Layout,
  Overflow,
  OverlayRegion,
  Position,
  Radio,
  TransitionGroup,
  TransitionType,
  ZIndex,
} from "v2";
import { Component, FormEvent } from "react";
import { Padding, cn, newUUIDv4 } from "@uuixjs/uuixweb-lib";

import { ExampleThumbnails } from "../../assets";

export enum ExampleBG {
  Off = "non-overlay",
  Photo1 = "photo-1",
  Photo2 = "photo-2",
  Video = "video",
  Black = "black",
  White = "white",
}

interface Props {
  defaultBg?: ExampleBG;
  blur?: boolean;
  padding?: Padding;
}

interface State {
  bg: ExampleBG;
}

export class OverlayPreview extends Component<Props, State> {
  public static defaultProps = {
    padding: 2,
    blur: true,
  };

  private uid = newUUIDv4();

  public constructor(props: Props) {
    super(props);
    this.state = {
      bg: props.defaultBg || ExampleBG.Black,
    };
  }

  public render() {
    const overlay = this.state.bg !== ExampleBG.Off;

    return (
      <Layout>
        <Layout display={Display.Flex} padding={1}>
          {Object.keys(ExampleBG).map((k, i) => (
            <Layout key={k} padding={{ left: i > 0 ? 1 : 0 }}>
              <Radio
                name={`overlay-preview-${this.uid}`}
                label={k}
                value={ExampleBG[k as keyof typeof ExampleBG]}
                onChange={this.onRadioSelected}
                checked={
                  this.state.bg === ExampleBG[k as keyof typeof ExampleBG]
                }
              />
            </Layout>
          ))}
        </Layout>

        <OverlayRegion overlay={overlay}>
          <Layout
            className="overlay-preview__container"
            position={Position.Relative}
            background={overlay ? Background.Overlay : undefined}
            padding={this.props.padding}
          >
            {this.renderBackground()}
            <Layout zIndex={ZIndex.Above} position={Position.Relative}>
              {this.props.children}
            </Layout>
          </Layout>
        </OverlayRegion>
      </Layout>
    );
  }

  private renderBackground() {
    let element;
    if (this.state.bg === ExampleBG.Photo1) {
      element = <CoreImage src={ExampleThumbnails.stream1} alt="" />;
    }

    if (this.state.bg === ExampleBG.Photo2) {
      element = <CoreImage src={ExampleThumbnails.stream2} alt="" />;
    }

    if (this.state.bg === ExampleBG.Video) {
      element = (
        <video
          src="https://d3hmvhl7ru3t12.cloudfront.net/video/pages/home/header-82b9abce067992beedde6d103b6241d616a7ee97cc0d4f3126277098c2e4b09004325f25c6a245656096bd333c880b0032e2cbdbed9d2514220e20718482e95b.mp4"
          autoPlay
          muted
          loop
        />
      );
    }

    if (this.state.bg === ExampleBG.Black) {
      element = <div style={{ background: "#000" }} />;
    }

    if (this.state.bg === ExampleBG.White) {
      element = <div style={{ background: "#FFF" }} />;
    }

    return (
      <InjectLayout
        display={Display.Flex}
        alignItems={AlignItems.Stretch}
        attachTop
        attachLeft
        attachBottom
        attachRight
        overflow={Overflow.Hidden}
      >
        <TransitionGroup
          className="overlay-preview__transition-group-container"
          transitionType={TransitionType.Fade}
        >
          {element ? (
            <InjectLayout
              className={cn({
                ["overlay-preview__bg-element"]: true,
                ["overlay-preview__bg-element--blur"]: this.props.blur,
              })}
              key={this.state.bg}
              position={Position.Absolute}
              attachTop
              attachLeft
              attachBottom
              attachRight
              fullWidth
              fullHeight
            >
              {element}
            </InjectLayout>
          ) : (
            []
          )}
        </TransitionGroup>
      </InjectLayout>
    );
  }

  private onRadioSelected = (e: FormEvent<HTMLInputElement>) => {
    if (e.currentTarget.checked) {
      this.setState({
        bg: e.currentTarget.value as ExampleBG,
      });
    }
  };
}
