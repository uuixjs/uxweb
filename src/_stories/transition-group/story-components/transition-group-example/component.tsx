import * as React from "react";

import {
  AlignItems,
  Background,
  Button,
  ButtonType,
  Column,
  Display,
  FlexDirection,
  FlexWrap,
  FormGroup,
  Grid,
  Input,
  InputType,
  JustifyContent,
  Layout,
  Range,
  Select,
  Tower,
  TowerChildWidth,
  TowerGutter,
  TransitionDuration,
  TransitionGroup,
  TransitionType,
  TransitionTypeOption,
} from "v2";
import { Component, FormEvent, MouseEvent } from "react";

import { TransitionGroupChecklist } from "../transition-group-checklist";
import { newUUIDv4 } from "@uuixjs/uuixweb-lib";

enum DemoLayoutType {
  Ordinary = "ordinary",
  Flexbox = "flexbox",
  // eslint-disable-next-line @typescript-eslint/no-shadow
  Tower = "tower",
}

interface Item {
  key: string;
}

export interface UpdateSettings {
  numToAdd: number;
  addAtIndex: number;
  numToRemove: number;
  removeAtIndex: number;
  numToMove: number;
  moveFromIndex: number;
  moveToIndex: number;
}

export interface LayoutSettings {
  layout: DemoLayoutType;
  direction: FlexDirection;
  duration: TransitionDuration | number;
  transitionType: TransitionTypeOption;
  boxSize: number;
}

interface Props {}

interface State extends UpdateSettings, LayoutSettings {
  showItems: boolean;
  items: Item[];
}

const INITIAL_VALUES: UpdateSettings = {
  numToAdd: 1,
  addAtIndex: 1,
  numToRemove: 0,
  removeAtIndex: 1,
  numToMove: 0,
  moveFromIndex: 0,
  moveToIndex: 1,
};

export class TransitionGroupExample extends Component<Props, State> {
  public state: State = {
    ...INITIAL_VALUES,
    layout: DemoLayoutType.Ordinary,
    direction: FlexDirection.Column,
    duration: TransitionDuration.Long,
    transitionType: [TransitionType.ScaleOver, TransitionType.ScaleOver],
    boxSize: 240,
    showItems: true,
    items: [],
  };

  public componentDidMount() {
    this.init();
  }

  public render() {
    return (
      <Layout margin={{ top: 2 }} border background={Background.Base}>
        <Grid>
          <Column cols={{ default: 12, md: 6, lg: 5, xxl: 4 }}>
            {this.renderListControls()}
          </Column>

          <Column cols={{ default: 12, md: 6, lg: 7, xxl: 8 }}>
            {this.renderLayoutControls()}
          </Column>

          <Column cols={{ default: 12, lg: 12, xxl: 8 }}>
            {this.state.showItems && (
              <div
                style={{
                  display: "block",
                  border: "2px dashed lightgray",
                  margin: "2rem",
                }}
              >
                {this.renderExample()}
              </div>
            )}
          </Column>

          <Column cols={{ default: 12, lg: 12, xxl: 4 }}>
            <Layout padding={2}>
              <TransitionGroupChecklist
                setValues={this.setValues}
                updateView={this.changeItems}
                setItemCount={this.resetItemsWithoutTransition}
                addItem={this.addItem}
                removeItem={this.removeItem}
              />
            </Layout>
          </Column>
        </Grid>
      </Layout>
    );
  }

  private renderExample() {
    if (this.state.layout === DemoLayoutType.Ordinary) {
      return (
        <TransitionGroup
          duration={this.state.duration}
          transitionType={this.state.transitionType}
        >
          {this.renderItems()}
        </TransitionGroup>
      );
    }
    if (this.state.layout === DemoLayoutType.Flexbox) {
      return (
        <TransitionGroup
          duration={this.state.duration}
          transitionType={this.state.transitionType}
          component={
            <Layout
              display={Display.Flex}
              flexDirection={this.state.direction}
              flexWrap={FlexWrap.Wrap}
              padding={1}
            />
          }
        >
          {this.renderItems()}
        </TransitionGroup>
      );
    }

    if (this.state.layout === DemoLayoutType.Tower) {
      return (
        <TransitionGroup
          duration={this.state.duration}
          transitionType={this.state.transitionType}
          component={
            <Tower
              childWidth={TowerChildWidth.Medium}
              gutterSize={TowerGutter.ExtraSmall}
              placeholderItems={20}
            />
          }
        >
          {this.renderItems()}
        </TransitionGroup>
      );
    }
  }

  private renderItems() {
    return this.state.items.map((item, index) => (
      <div
        key={item.key}
        style={{
          display: "flex",
          flexDirection: "column",
          justifyContent: "center",
          alignItems: "center",
          width: `100%`,
          height: this.state.boxSize ? `${this.state.boxSize / 2}px` : "100px",
          background: "green",
          color: "rgba(255,255,255,0.9)",
          fontSize: "1.5rem",
          border: "2px solid darkgreen",
          margin: "",
        }}
      >
        <Layout
          display={Display.Flex}
          flexDirection={FlexDirection.Column}
          alignItems={AlignItems.Center}
          margin={{ bottom: 1 }}
        >
          <div>index: {index}</div>
          <Layout display={Display.Flex} alignItems={AlignItems.Center}>
            <div>key: {item.key.substring(item.key.length - 6)}</div>
            <div
              style={{
                width: "20px",
                height: "20px",
                background: `#${item.key.substring(item.key.length - 6)}`,
                margin: "0 0.5rem",
              }}
            />
          </Layout>
        </Layout>
        <Button
          variant={ButtonType.Secondary}
          onClick={() => this.removeItem(item.key)}
          overlay
        >
          Remove
        </Button>
      </div>
    ));
  }

  private renderListControls = () => {
    return (
      <form onSubmit={this.onSubmitForm}>
        <Layout
          display={Display.InlineFlex}
          flexDirection={FlexDirection.Column}
          alignItems={AlignItems.End}
          padding={2}
        >
          <Layout margin={{ bottom: 1 }}>
            Move
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.numToMove.toString()}
                onChange={this.setNumToMove}
                min={0}
                max={10}
              />
            </div>
            items from
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.moveFromIndex.toString()}
                onChange={this.setMoveFromIndex}
                min={0}
              />
            </div>
            to
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0 0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.moveToIndex.toString()}
                onChange={this.setMoveToIndex}
                min={0}
              />
            </div>
          </Layout>

          <Layout margin={{ top: 0.5 }}>
            Add
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.numToAdd.toString()}
                onChange={this.setNumToAdd}
                min={0}
                max={10}
              />
            </div>
            items at index
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0 0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.addAtIndex.toString()}
                onChange={this.setAddAtIndex}
                min={0}
              />
            </div>
          </Layout>

          <Layout margin={{ top: 0.5 }}>
            Remove
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.numToRemove.toString()}
                onChange={this.setNumToRemove}
                min={0}
                max={10}
              />
            </div>
            items at index
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0 0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.removeAtIndex.toString()}
                onChange={this.setRemoveAtIndex}
                min={0}
              />
            </div>
          </Layout>

          <Layout display={Display.Flex} margin={{ top: 1 }}>
            <Layout margin={{ right: 1 }}>
              <Button onClick={this.onSubmitForm}>Update</Button>
            </Layout>
            <Button onClick={this.init} variant={ButtonType.Secondary}>
              Reset
            </Button>
          </Layout>
        </Layout>
      </form>
    );
  };

  private generateItem = (): Item => {
    return {
      key: newUUIDv4(),
    };
  };

  private init = (e?: MouseEvent<HTMLElement>) => {
    if (e) {
      e.preventDefault();
    }

    this.setState({
      ...INITIAL_VALUES,
      // Initial items have static keys and color values to ensure screenshot diff passes
      items: [{ key: "aa0000" }, { key: "bbbb00" }, { key: "0033cc" }],
    });
  };

  private resetItemsWithoutTransition = async (num: number) => {
    return new Promise<void>((resolve) => {
      this.setState(
        {
          showItems: false,
        },
        () => {
          // wait for full re-render
          requestAnimationFrame(() => {
            this.setState(
              {
                showItems: true,
                items: Array.from(Array(num), this.generateItem),
              },
              resolve,
            );
          });
        },
      );
    });
  };

  private onSubmitForm = (
    e: FormEvent<HTMLFormElement> | MouseEvent<HTMLElement>,
  ) => {
    e.preventDefault();
    this.changeItems();
  };

  private changeItems = () => {
    this.setState((prevState) => {
      const {
        numToAdd,
        addAtIndex,
        numToRemove,
        removeAtIndex,
        numToMove,
        moveFromIndex,
        moveToIndex,
      } = prevState;

      const items = prevState.items.concat();
      const newItems = Array.from(Array(numToAdd), this.generateItem);

      // Remove some
      items.splice(removeAtIndex, numToRemove);

      // Add some
      items.splice(addAtIndex, 0, ...newItems);

      // Move some
      const moving = items.splice(moveFromIndex, numToMove);
      items.splice(moveToIndex, 0, ...moving);

      return {
        items,
      };
    });
  };

  private addItem = (key: string, index: number = 0) => {
    this.setState((prevState) => {
      const items = prevState.items.concat();
      items.splice(index, 0, { key });
      return {
        items,
      };
    });
  };

  private removeItem = (key: string) => {
    this.setState((prevState) => ({
      items: prevState.items.filter((item) => item.key !== key),
    }));
  };

  private setNumToAdd = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      numToAdd: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setAddAtIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      addAtIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setNumToRemove = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      numToRemove: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setRemoveAtIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      removeAtIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setNumToMove = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      numToMove: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setMoveFromIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      moveFromIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setMoveToIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      moveToIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setValues = async (settings?: Partial<UpdateSettings>) => {
    return new Promise<void>((resolve) =>
      this.setState(
        {
          ...INITIAL_VALUES,
          ...settings,
        },
        resolve,
      ),
    );
  };

  private renderLayoutControls() {
    return (
      <Layout
        padding={2}
        fullWidth
        display={Display.Flex}
        justifyContent={JustifyContent.Start}
        flexWrap={FlexWrap.Wrap}
      >
        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="Layout Demo">
            <Select value={this.state.layout} onChange={this.updateLayout}>
              {Object.keys(DemoLayoutType).map((key: string) => (
                <option
                  value={DemoLayoutType[key as keyof typeof DemoLayoutType]}
                  key={DemoLayoutType[key as keyof typeof DemoLayoutType]}
                >
                  {key}
                </option>
              ))}
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="FlexDirection">
            <Select
              value={this.state.direction}
              onChange={this.updateDirection}
              disabled
            >
              <option value={FlexDirection.Row}>Row</option>
              <option value={FlexDirection.Column}>Column</option>
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="TransitionDuration">
            <Select
              value={String(this.state.duration)}
              onChange={this.updateDuration}
            >
              {Object.keys(TransitionDuration).map((key: string) => (
                <option
                  value={
                    TransitionDuration[key as keyof typeof TransitionDuration]
                  }
                  key={
                    TransitionDuration[key as keyof typeof TransitionDuration]
                  }
                >
                  {key}
                </option>
              ))}
              <option value="1000" key="1000ms">
                1,000ms
              </option>
              <option value="3000" key="3000ms">
                3,000ms
              </option>
              <option value="5000" key="5000ms">
                5,000ms
              </option>
              <option value="10000" key="10000ms">
                10,000ms
              </option>
              <option value="0" key="0ms">
                0ms
              </option>
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="TransitionType [Enter]">
            <Select
              value={this.state.transitionType[0]}
              onChange={this.updateTransitionTypeEnter}
            >
              {Object.keys(TransitionType).map((key: string) => (
                <option
                  value={TransitionType[key as keyof typeof TransitionType]}
                  key={TransitionType[key as keyof typeof TransitionType]}
                >
                  {key}
                </option>
              ))}
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="TransitionType [Exit]">
            <Select
              value={this.state.transitionType[1]}
              onChange={this.updateTransitionTypeExit}
            >
              {Object.keys(TransitionType).map((key: string) => (
                <option
                  value={TransitionType[key as keyof typeof TransitionType]}
                  key={TransitionType[key as keyof typeof TransitionType]}
                >
                  {key}
                </option>
              ))}
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="Box Size">
            <Range
              min={100}
              max={600}
              value={this.state.boxSize.toString()}
              onChange={this.updateBoxSize}
              disabled
            />
          </FormGroup>
        </Layout>
      </Layout>
    );
  }

  private updateLayout = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      layout: e.currentTarget.value as DemoLayoutType,
    });
  };

  private updateDirection = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      direction: e.currentTarget.value as FlexDirection,
    });
  };

  private updateDuration = (e: FormEvent<HTMLSelectElement>) => {
    if (e.currentTarget.value.match(/^[0-9]+$/)) {
      this.setState({
        duration: Number(e.currentTarget.value),
      });
    } else {
      this.setState({
        duration: e.currentTarget.value as TransitionDuration,
      });
    }
  };

  private updateTransitionTypeEnter = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      transitionType: [
        e.currentTarget.value as TransitionType,
        this.state.transitionType[1] as TransitionType,
      ],
    });
  };

  private updateTransitionTypeExit = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      transitionType: [
        this.state.transitionType[0] as TransitionType,
        e.currentTarget.value as TransitionType,
      ],
    });
  };

  private updateBoxSize = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      boxSize: Number.parseInt(e.currentTarget.value, undefined),
    });
  };
}
