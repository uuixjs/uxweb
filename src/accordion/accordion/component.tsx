import { AccordionBody, AccordionBodyProps } from "../accordion-body";
import { AccordionFooter, AccordionFooterProps } from "../accordion-footer";
import { AccordionHeader, AccordionHeaderProps } from "../accordion-header";
import { BorderRadius, getAriaProps } from "@uuixjs/uuixweb-lib";
import { BorderRadiusProps, Layout } from "../../layout";
import { Component, MouseEvent } from "react";

export type AccordionItem = {
  header: AccordionHeaderProps;
  body: AccordionBodyProps;
  footer?: AccordionFooterProps;
};

export interface AccordionProps {
  items: AccordionItem[];
  initialOpenIndex?: number | false;
}

export interface AccordionState {
  currentOpenIndex: number | false;
}

/**
 * Accordion is used to consistently compose and control some accordion items
 * which will each be composed of an AccordionHeader, AccordionBody, and AccordionFooter
 */
export class Accordion extends Component<AccordionProps, AccordionState> {
  public state: AccordionState = {
    currentOpenIndex:
      this.props.initialOpenIndex !== undefined
        ? this.props.initialOpenIndex
        : 0,
  };

  public render() {
    return (
      <Layout {...getAriaProps(this.props)}>
        {this.props.items.map((item, index) => (
          <Layout
            borderBottom={index !== this.props.items.length - 1}
            key={index}
          >
            <AccordionHeader
              {...item.header}
              isOpen={index === this.state.currentOpenIndex}
              onClick={this.onHeaderClick}
              index={index}
              borderRadius={this.getHeaderBorderRadius(index)}
            />

            <AccordionBody
              {...item.body}
              isOpen={index === this.state.currentOpenIndex}
              borderRadius={this.getBodyBorderRadius(index)}
            >
              {item.body.children}
              {item.footer && <AccordionFooter {...item.footer} />}
            </AccordionBody>
          </Layout>
        ))}
      </Layout>
    );
  }

  private onHeaderClick = (_EVENT: MouseEvent<HTMLElement>, index?: number) => {
    if (typeof index === "number") {
      this.setState((prevState) => ({
        currentOpenIndex: prevState.currentOpenIndex === index ? false : index,
      }));
    }
  };

  private getHeaderBorderRadius = (
    index: number,
  ): BorderRadiusProps | undefined => {
    const corners: BorderRadiusProps = {};

    if (index === 0) {
      corners.topLeft = BorderRadius.Medium;
      corners.topRight = BorderRadius.Medium;
    }

    if (
      index === this.props.items.length - 1 &&
      this.state.currentOpenIndex !== index
    ) {
      corners.bottomLeft = BorderRadius.Medium;
      corners.bottomRight = BorderRadius.Medium;
    }

    return corners;
  };

  private getBodyBorderRadius = (
    index: number,
  ): BorderRadiusProps | undefined => {
    if (index === this.props.items.length - 1) {
      return {
        bottomLeft: BorderRadius.Medium,
        bottomRight: BorderRadius.Medium,
      };
    }

    return undefined;
  };
}
