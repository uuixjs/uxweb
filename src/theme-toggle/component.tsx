import {
  FormEventHandler,
  ForwardRefRenderFunction,
  Ref,
  forwardRef,
} from "react";
import {
  css,
  getAriaProps,
  getDataProps,
  hoverCss,
  newUUIDv4,
  staticTokenRule,
  styleVariant,
  styled,
} from "@uuixjs/uuixweb-lib";

import { CoreFocusEventHandlers } from "../core-interactive";
import { darken } from "polished";
import { focusRingStyleReplacement } from "../form/common";

export interface ThemeToggleProps
  extends CoreFocusEventHandlers<HTMLInputElement> {
  checked?: boolean;
  defaultValue?: string;
  disabled?: boolean;
  id?: string;
  name?: string;
  onChange?: FormEventHandler<HTMLInputElement>;
  /** @deprecated Ref forwarding is now supported; use the `ref` prop instead of `refDelegate` */
  refDelegate?: Ref<HTMLInputElement>;
  tabIndex?: number;
  value?: string;
}

// Dark mode toggle with sun and moon
const boxShadowColor = "rgba(0, 0, 0, 0.1)";
const lightToggle = "#ffc361";
const darkToggle = "#fffbf3";

const darkToggleBackground = "#6d51aa";
const darkToggleBackgroundHover = staticTokenRule("color-twitch-purple-7");
const lightToggleBackground = "#8eb4e0";
const lightToggleBackgroundHover = darken(0.1, lightToggleBackground);

const crater = "#d8d3cb";
const glare = "#ffe1a9";
const rayHeight = "2px";

const time = staticTokenRule("timing-short");

const height = "25px";
const width = "55px";
const offset = "6px";

const ScThemeToggleItems = styled.span`
  background-color: ${lightToggle};
  border-radius: 50%;
  box-shadow: 0 2px 4px ${boxShadowColor};
  display: inline-block;
  height: calc(${height} - ${offset});
  left: calc(${offset} / 2);
  position: relative;
  top: calc(${offset} / 2);
  transform: rotate(-90deg);
  transition-duration: ${time};
  transition-property: transform, top, left;
  width: calc(${height} - ${offset});
`;

const ScThemeToggleGlare = styled.span`
  border: solid calc(${height} / 14) ${glare};
  border-color: transparent transparent ${glare};
  border-radius: 0 0 200px 200px;
  height: calc(${height} / 2);
  left: 15%;
  opacity: 1;
  position: absolute;
  top: 15%;
  transform: rotate(220deg);
  transition: opacity calc(${time} / 4);
  width: calc(${height} / 2);
`;

const ScThemeToggleDot = styled.span<{ dot: 1 | 2 | 3 }>`
  background-color: ${crater};
  border-radius: 100%;
  opacity: 0;
  position: absolute;
  transition: opacity calc(${time} / 4);

  ${styleVariant("dot", {
    1: css`
      height: calc(${height} / 7);
      left: 25%;
      top: 30%;
      width: calc(${height} / 7);
    `,
    2: css`
      height: calc(${height} / 6);
      left: 50%;
      top: 60%;
      width: calc(${height} / 6);
    `,
    3: css`
      height: calc(${height} / 5);
      right: 20%;
      top: 20%;
      width: calc(${height} / 5);
    `,
  })}
`;

const ScThemeToggleRays = styled.span`
  opacity: 1;
  pointer-events: none;
  transition: opacity calc(${time} / 4);
  transition-delay: calc(${time} / 4);
`;

const ScThemeToggleRay = styled.span<{ ray: 1 | 2 | 3 }>`
  background: ${lightToggle};
  border-radius: ${offset};
  height: ${rayHeight};
  position: absolute;

  ${styleVariant("ray", {
    1: css`
      width: calc(${height} / 4);
      left: calc(${height} + 1px);
      top: calc(${height} / 2 - ${offset} / 4);
    `,
    2: css`
      left: ${height};
      top: calc(20%);
      transform: rotate(-20deg);
      width: calc(${height} / 5);
    `,
    3: css`
      left: ${height};
      top: calc(80% - 3px);
      transform: rotate(20deg);
      width: calc(${height} / 5);
    `,
  })}
`;

const ScThemeToggleInput = styled.input`
  position: absolute;
  opacity: 0;
`;

const ScThemeToggleButton = styled.label`
  background-color: ${lightToggleBackground};
  border-radius: calc(${width} - ${offset});
  cursor: pointer;
  display: inline-block;
  height: ${height};
  position: relative;
  transition: background-color ${time};
  width: ${width};

  ${focusRingStyleReplacement(ScThemeToggleInput)}

  ${hoverCss`
    background: ${lightToggleBackgroundHover};
  `}

  ${ScThemeToggleInput} {
    + && {
      opacity: 0.5;
      pointer-events: none;
    }
  }

  ${ScThemeToggleInput}:checked + && {
    background-color: ${darkToggleBackground};

    ${hoverCss`
      background: ${darkToggleBackgroundHover};
    `}

    // Need to set the ray opacity to 0 for Edge, otherwise they are visible. Changed from the container
    ${ScThemeToggleRay} {
      opacity: 0;
    }

    ${ScThemeToggleItems} {
      background-color: ${darkToggle};
      transform: translate3d(calc(${width} - ${height}), 0, 0) rotate(0);

      ${ScThemeToggleDot} {
        opacity: 1;
      }

      ${ScThemeToggleGlare} {
        opacity: 0;
      }
    }
  }
`;

const ScThemeToggle = styled.div`
  position: relative;
  user-select: none;
`;

const ThemeToggle: ForwardRefRenderFunction<
  HTMLInputElement,
  ThemeToggleProps
> = (props, ref) => {
  const id = props.id || newUUIDv4();

  return (
    <ScThemeToggle className="tw-theme-toggle" {...getDataProps(props)}>
      <ScThemeToggleInput
        className="tw-theme-toggle__input"
        ref={ref || props.refDelegate}
        checked={props.checked}
        defaultValue={props.defaultValue}
        disabled={props.disabled}
        id={id}
        name={props.name}
        onBlur={props.onBlur}
        onChange={props.onChange}
        onFocus={props.onFocus}
        tabIndex={props.tabIndex}
        type="checkbox"
        value={props.value}
        data-a-target="tw-theme-toggle"
        {...getAriaProps(props)}
      />
      <ScThemeToggleButton htmlFor={id}>
        <ScThemeToggleRays>
          <ScThemeToggleRay ray={1} />
          <ScThemeToggleRay ray={2} />
          <ScThemeToggleRay ray={3} />
        </ScThemeToggleRays>
        <ScThemeToggleItems>
          <ScThemeToggleGlare />
          <ScThemeToggleDot dot={1} />
          <ScThemeToggleDot dot={2} />
          <ScThemeToggleDot dot={3} />
        </ScThemeToggleItems>
      </ScThemeToggleButton>
    </ScThemeToggle>
  );
};

ThemeToggle.displayName = "ThemeToggle";
const ComponentWithRef = forwardRef(ThemeToggle);
export { ComponentWithRef as ThemeToggle };
