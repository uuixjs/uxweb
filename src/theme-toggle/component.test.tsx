import { setupShallowTest } from "../../tests/helpers";
import { ThemeToggle, ThemeToggleProps } from "./component";

function getOptionalProps(): ThemeToggleProps {
  return {
    checked: true,
    defaultValue: "Default value",
    disabled: true,
    id: "ThemeToggle test ID",
    name: "ThemeToggle test name",
    onBlur: jest.fn(),
    onChange: jest.fn(),
    onFocus: jest.fn(),
    tabIndex: 0,
    value: "ThemeToggle test value",
  };
}

const setupShallow = setupShallowTest(ThemeToggle);

describe("ThemeToggle", () => {
  it("renders theme toggle component with required props", () => {
    const { wrapper } = setupShallow({ id: "ThemeToggle test ID" });
    expect(wrapper).toMatchSnapshot();
  });

  it("renders theme toggle component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });
});
