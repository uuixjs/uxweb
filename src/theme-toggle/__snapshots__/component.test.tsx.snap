// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeToggle renders theme toggle component with all props 1`] = `
<styled.div
  className="tw-theme-toggle"
>
  <styled.input
    checked={true}
    className="tw-theme-toggle__input"
    data-a-target="tw-theme-toggle"
    defaultValue="Default value"
    disabled={true}
    id="ThemeToggle test ID"
    name="ThemeToggle test name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    tabIndex={0}
    type="checkbox"
    value="ThemeToggle test value"
  />
  <styled.label
    htmlFor="ThemeToggle test ID"
  >
    <styled.span>
      <styled.span
        ray={1}
      />
      <styled.span
        ray={2}
      />
      <styled.span
        ray={3}
      />
    </styled.span>
    <styled.span>
      <styled.span />
      <styled.span
        dot={1}
      />
      <styled.span
        dot={2}
      />
      <styled.span
        dot={3}
      />
    </styled.span>
  </styled.label>
</styled.div>
`;

exports[`ThemeToggle renders theme toggle component with required props 1`] = `
<styled.div
  className="tw-theme-toggle"
>
  <styled.input
    className="tw-theme-toggle__input"
    data-a-target="tw-theme-toggle"
    id="ThemeToggle test ID"
    type="checkbox"
  />
  <styled.label
    htmlFor="ThemeToggle test ID"
  >
    <styled.span>
      <styled.span
        ray={1}
      />
      <styled.span
        ray={2}
      />
      <styled.span
        ray={3}
      />
    </styled.span>
    <styled.span>
      <styled.span />
      <styled.span
        dot={1}
      />
      <styled.span
        dot={2}
      />
      <styled.span
        dot={3}
      />
    </styled.span>
  </styled.label>
</styled.div>
`;
